### PR TITLE
RSDK-5161 make Navigation service use MoveOnGlobeNew

### PR DESCRIPTION
**Pull request by nicksanford**
Tickets:
https://viam.atlassian.net/browse/RSDK-5161
https://viam.atlassian.net/browse/RSDK-4819
https://viam.atlassian.net/browse/RSDK-5193
https://viam.atlassian.net/browse/RSDK-5525

- Change `nav.SetMode` to call `motion.MoveOnGlobeNew` rather than `motion.MoveOnGlobe` & poll `motion.PlanHistory` until the plan terminates, retrying the waypoint on failure & moving on to the next waypoint on success. Also (for use by nav.Paths) record the executionID & waypoint of the in progress moveOnGlobe call & unset it when moveOnGlobe terminates. 
- Change nav to always set position only mode on motion
- Uncomment the tests cases for `nav.SetMode` that affect waypoints & fix the tests to work with `motion.MoveOnGlobeNew`
- Implement `nav.Paths`
- Change MoveOnGlobeNew's go interface method to return a motion.ExecutionID rather than a string to improve ergonomics.
- Change `nav.Path`'s destinationWaypointID field to be of type primitive.ObjectID rather than string

Manual Testing: In gazebo:
https://github.com/viamrobotics/rdk/assets/5927876/3994873f-18dd-412a-9654-96b31d2327f0

Script used in manual testing:
```go
package main

import (
	"context"
	"fmt"

	"github.com/edaniels/golog"
	"go.viam.com/rdk/robot/client"
	"go.viam.com/rdk/services/navigation"
	"go.viam.com/utils/rpc"
)

func main() {
	logger := golog.NewDevelopmentLogger("client")

	robot, err := client.New(
		context.Background(),
		"",
		logger,
		client.WithDialOptions(rpc.WithEntityCredentials(
			"",
			rpc.Credentials{
				Type:    rpc.CredentialsTypeAPIKey,
				Payload: "",
			})),
	)
	if err != nil {
		logger.Fatal(err)
	}

	defer robot.Close(context.Background())

	n, err := navigation.FromRobot(robot, "nav")
	if err != nil {
		logger.Error(err)
		return
	}

	resp, err := n.Paths(context.Background(), nil)
	if err != nil {
		logger.Error(err)
		return
	}

	for _, r := range resp {
		fmt.Printf("WaypointID: %s\n", r.DestinationWaypointID())
		fmt.Print("GeoPoints:")
		for _, gp := range r.GeoPoints() {
			fmt.Printf("Lat: %f, Lng: %f\n", gp.Lat(), gp.Lng())
		}

	}
}
```

### CODE REVIEW

File: services/navigation/builtin/builtin.go [hunk 1]
```diff
@@ -539,6 +596,7 @@ func (svc *builtIn) startWaypointMode(ctx context.Context, extra map[string]inte
 
 			wp, err := svc.store.NextWaypoint(ctx)
 			if err != nil {
+				time.Sleep(time.Millisecond * 50)
```
Comment (nicksanford) [inline services/navigation/builtin/builtin.go:599]: This sleep is to limit busy waiting, not required for correctness but just prevents wasting CPU

File: services/navigation/builtin/builtin.go [hunk 2]
```diff
@@ -206,12 +219,13 @@ func NewBuiltIn(
 
 type builtIn struct {
 	resource.Named
-	actionMu  sync.RWMutex
-	mu        sync.RWMutex
-	store     navigation.NavStore
-	storeType string
-	mode      navigation.Mode
-	mapType   navigation.MapType
+	activeExecutionWaypoint atomic.Value
```
Comment (nicksanford) [inline services/navigation/builtin/builtin.go:222]: `activeExecutionWaypoint` is added as an atomic value so it can be updated across threads without needing to take a lock on `mu`, which was useful in preventing unnecessary race conditions & deadlocks.

File: services/navigation/builtin/builtin.go [hunk 3]
```diff
@@ -495,42 +509,86 @@ func (svc *builtIn) Close(ctx context.Context) error {
...
+	var stopWG sync.WaitGroup
+	defer stopWG.Wait()
+	cancelCtx, cancelFn := context.WithCancel(ctx)
+	defer cancelFn()
+	rawExecutionID, err := svc.motionService.MoveOnGlobeNew(cancelCtx, req)
+	if err != nil {
+		return err
+	}
+
+	executionID, err := uuid.Parse(rawExecutionID)
```
Comment (raybjork) [inline services/navigation/builtin/builtin.go:529]: Why is this necessary?

File: services/navigation/builtin/builtin.go [hunk 4]
```diff
@@ -495,42 +509,86 @@ func (svc *builtIn) Close(ctx context.Context) error {
...
+	if err != nil {
+		return err
+	}
+	executionWaypoint := executionWaypoint{executionID: executionID, waypoint: wp}
+	if old := svc.activeExecutionWaypoint.Swap(executionWaypoint); old != nil && old != emptyExecutionWaypoint {
```
Comment (raybjork) [inline services/navigation/builtin/builtin.go:534]: This is a good idea.  Do we need to check this part though? `old != emptyExecutionWaypoint`

File: services/navigation/builtin/builtin.go [hunk 5]
```diff
@@ -495,42 +509,86 @@ func (svc *builtIn) Close(ctx context.Context) error {
...
+		msg := "unexpected race condition in moveOnGlobeSync, expected " +
+			"replaced waypoint & execution id to be nil or %#v; instead was %s"
+		svc.logger.Errorf(msg, emptyExecutionWaypoint, old)
+	}
+	stopWG.Add(1)
+	// call StopPlan upon exiting moveOnGlobeSync
+	// is a NoOp if execution has already terminted
+	utils.ManagedGo(func() {
+		<-cancelCtx.Done()
```
Comment (raybjork) [inline services/navigation/builtin/builtin.go]: I dont think this has to be its own thing in a goroutine and that we can just call stop when we need to stop

File: services/navigation/builtin/builtin.go [hunk 6]
```diff
@@ -581,5 +642,90 @@ func (svc *builtIn) Obstacles(ctx context.Context, extra map[string]interface{})
 func (svc *builtIn) Paths(ctx context.Context, extra map[string]interface{}) ([]*navigation.Path, error) {
 	svc.mu.RLock()
 	defer svc.mu.RUnlock()
-	return nil, errUnimplemented
+
+	rawExecutionWaypoint := svc.activeExecutionWaypoint.Load()
+	// If there is no execution, return empty paths
+	if rawExecutionWaypoint == nil || rawExecutionWaypoint == emptyExecutionWaypoint {
```
Comment (raybjork) [inline services/navigation/builtin/builtin.go:644]: I don't think we need the second check here we should be able to control what is put into an executionWaypoint and this should never come up

File: services/navigation/builtin/builtin.go [hunk 7]
```diff
@@ -581,5 +642,90 @@ func (svc *builtIn) Obstacles(ctx context.Context, extra map[string]interface{})
...
+		err := errors.New("plan failed")
+		if reason := status.Reason; reason != nil {
+			err = errors.New(*reason)
+		}
+		return err
+
+	case motion.PlanStateStopped:
+		return errors.New("plan stopped")
+
+	case motion.PlanStateInProgress, motion.PlanStateSucceeded:
```
Comment (raybjork) [inline services/navigation/builtin/builtin.go:695]: Why break this into its own function?  Seems a little redundant since you still have to check the good cases too

**Review by raybjork — CHANGES_REQUESTED** [review]
I haven't looked at the test code yet but looks pretty good so far.  A few comments

File: services/navigation/builtin/builtin.go (hunk 1 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:596]: They both happen to be 50 ms but are not directly coupled. I've made another constant for this sleep frequency.

File: services/navigation/builtin/builtin.go (hunk 3 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:529]: b/c `MoveOnGlobeNew`returns a `string,error`, where the string is a stringified executionID, which is a uuid, but `motion.PlanHistory` takes the a non-stringified executionID. Hence there is need to parse the rawExeuctionID (string) into a uuid.

File: services/navigation/builtin/builtin.go (hunk 4 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:534]: It is unfortunately. https://pkg.go.dev/sync/atomic#Value.Swap
from the docs:
```
Swap stores new into Value and returns the previous value. It returns nil if the Value is empty.

All calls to Swap for a given Value must use values of the same concrete type. Swap of an inconsistent type panics, as does Swap(nil).
```

So the first time we call `Swap` we will get back `nil`, but after that point it will never return nil, b/c if I try to call `Swap(nil)` it panics. That is why we set it to `emptyExecutionWaypoint` after calling `StopPlan`.

File: services/navigation/builtin/builtin.go (hunk 5 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go]: You are right. I've refactored it to be a defer

File: services/navigation/builtin/builtin.go (hunk 6 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:644]: The invariant we want to be maintained is that nav only has one Execution in progress at a time & never loses track of an execution it started.

In order for that invariant to be maintained, we need to ensure that we are only setting `activeExecutionWaypoint` when it is nil / empty & only unsetting it (setting it to empty) when it is set to the executionID we expected.

This checks those invariants & gives us feedback if someone in the future changes the code such that those invariants are not maintained.

 I can remove these checks as the current form of the code maintains those invariants & never goes down the sad paths if you would like me to, however, I would prefer keeping these checks so that a future developer that changes this code will get feedback if they change the code to violate the invariants.

File: services/navigation/builtin/builtin.go (hunk 3 above)
Comment (raybjork) [reply services/navigation/builtin/builtin.go:529]: Can't we just change MoveOnGlobeNew to return the executionID directly?  The client server code can handle the conversion to and from the string that it must exist as in proto

File: services/navigation/builtin/builtin.go (hunk 6 above)
Comment (raybjork) [reply services/navigation/builtin/builtin.go:644]: Ok that seems reasonable thanks for the exaplanation

File: services/navigation/builtin/builtin.go (hunk 1 above)
Comment (raybjork) [reply services/navigation/builtin/builtin.go:596]: In either case its the frequency at which we check for new plans correct?  In one case its the one where there is an error getting next waypoint (in practice because there are none) and in the other its that we're polling the history of a known waypoint?  I'd be ok collapsing them but this is fine too

File: services/navigation/builtin/builtin.go [hunk 8]
```diff
@@ -495,42 +509,86 @@ func (svc *builtIn) Close(ctx context.Context) error {
...
-		if false {
-			extra = map[string]interface{}{"motion_profile": "position_only"}
-		} // TODO: Fix with RSDK-4583
-	} else if _, ok := extra["motion_profile"]; !ok {
-		if false {
-			extra["motion_profile"] = "position_only"
-		} // TODO: Fix with RSDK-4583
+		extra = map[string]interface{}{
+			"motion_profile": "position_only",
+			"smooth_iter":    20,
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:574]: 👍 removed

File: services/navigation/builtin/builtin.go (hunk 1 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:596]: I've made them one frequency

File: services/navigation/builtin/builtin.go [hunk 9]
```diff
@@ -495,42 +509,86 @@ func (svc *builtIn) Close(ctx context.Context) error {
 	return svc.store.Close(ctx)
 }
 
+func (svc *builtIn) moveOnGlobeSync(ctx context.Context, wp navigation.Waypoint, extra map[string]interface{}) error {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:512]: done

File: services/navigation/builtin/builtin.go (hunk 3 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:529]: I'll get started on that now

Comment (nicksanford) [reply services/navigation/builtin/builtin.go]: done

File: services/motion/builtin/builtin_test.go [hunk 10]
```diff
@@ -639,7 +638,9 @@ func TestMoveOnMapSubsequent(t *testing.T) {
 		entry1GoalLine = strings.ReplaceAll(entry1GoalLine, ":", "\":")
 
 		posepb := &commonpb.Pose{}
-		json.Unmarshal([]byte(entry1GoalLine), posepb)
+		err := json.Unmarshal([]byte(entry1GoalLine), posepb)
+		test.That(t, err, test.ShouldBeNil)
```
Comment (nicksanford) [inline services/motion/builtin/builtin_test.go:642]: this was ignoring an error

File: services/navigation/builtin/builtin.go [hunk 11]
```diff
@@ -581,5 +624,81 @@ func (svc *builtIn) Obstacles(ctx context.Context, extra map[string]interface{})
...
+		var pose spatialmath.Pose
+		for _, p := range s {
+			pose = p
+		}
+
+		geoPoint := geo.NewPoint(pose.Point().Y, pose.Point().X)
+		geoPoints = append(geoPoints, geoPoint)
+	}
+
+	path, err := navigation.NewPath(ewp.waypoint.ID.String(), geoPoints)
```
Comment (raybjork) [inline services/navigation/builtin/builtin.go:662]: Seems like it would be better to keep this as they type of waypoint.ID rather than storing it as a string

File: services/navigation/builtin/builtin.go [hunk 12]
```diff
@@ -581,5 +624,81 @@ func (svc *builtIn) Obstacles(ctx context.Context, extra map[string]interface{})
...
+		}
+
+		status := ph[0].StatusHistory[0]
+
+		switch status.State {
+		case motion.PlanStateInProgress:
+		case motion.PlanStateFailed:
+			err := errors.New("plan failed")
+			if reason := status.Reason; reason != nil {
+				err = errors.New(*reason)
```
Comment (raybjork) [inline services/navigation/builtin/builtin.go:688]: Maybe we want to wrap the reason with the "plan failed"?  I'm not sure what form the reasons take so it might be that we already have this

**Review by raybjork — CHANGES_REQUESTED** [review]
I still need to read through half the builtin tests but in the meantime I had a few more nits

File: services/navigation/builtin/builtin.go [hunk 13]
```diff
@@ -581,5 +638,90 @@ func (svc *builtIn) Obstacles(ctx context.Context, extra map[string]interface{})
...
+		return nil, err
+	}
+
+	geoPoints := make([]*geo.Point, 0, len(ph[0].Plan.Steps))
+	for _, s := range ph[0].Plan.Steps {
+		if len(s) > 1 {
+			return nil, errors.New("multi component paths are unsupported")
+		}
+		var pose spatialmath.Pose
+		for _, p := range s {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/navigation/builtin/builtin.go:654]: `s` is a map which only has a single element. What is a better way to get the value of a single element map in go?

[the remark this answers is not in the document]
Comment (raybjork) [reply services/navigation/builtin/builtin.go:654]: Oh right I forgot it was a map and not a slice.

File: services/navigation/builtin/builtin.go (hunk 12 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go]: done

Comment (nicksanford) [reply services/navigation/builtin/builtin.go]: also it looks like this: `error state exceeds planDeviationMM; planDeviationMM: 1000.000000, errorstate.Point().Norm(): 1056.301307, errorstate.Point(): r3.Vector{X:-1037.7992823209859, Y:-196.83775392595362, Z:0} ` before the wrap

File: services/navigation/builtin/builtin.go (hunk 11 above)
Comment (nicksanford) [reply services/navigation/builtin/builtin.go]: done

**Review by raybjork — APPROVED** [review]
LGTM. Thanks!
